### PR TITLE
Feature/155 default unit of resource

### LINKED ISSUE #155: Add `ResourceSpecification.defaultUnitOfResource`
See https://github.com/valueflows/vf-graphql/pull/85

This is needed by rea-playspace project

### PR DESCRIPTION
**Pull request by Connoropolous**
fixes #155

### CODE REVIEW

File: modules/vf-graphql-holochain/resolvers/resourceSpecification.ts [hunk 1]
```diff
@@ -46,10 +46,11 @@ export default (enabledVFModules: VfModule[] = DEFAULT_VF_MODULES, dnaConfig: DN
       },
     } : {}),
     (hasMeasurement ? {
-      defaultUnitOfResource: () => {
-        // issue #155
-        // https://github.com/h-REA/hREA/issues/155
-        throw new Error('resolver unimplemented')
+      defaultUnitOfResource: async (record: { defaultUnitOfResource: AddressableIdentifier }) => {
+        if (!record.defaultUnitOfResource) {
```
Comment (pospi) [inline modules/vf-graphql-holochain/resolvers/resourceSpecification.ts:50]: can use the destructured property here, no?

File: zomes/rea_economic_resource/storage/src/lib.rs [hunk 2]
```diff
@@ -135,43 +135,70 @@ impl TryFrom<CreationPayload> for EntryData {
...
+                    None => default_unit_of_resource
+                };
+                Some(QuantityValue::new(resource_quantity.get_numerical_value(), unit))
+            },
+            _ => None,
+        };
+        // capture the unit_of_effort from any default
+        // set on the Resource Specification this conforms to,
+        // if any
+        let unit_of_effort = match maybe_resource_specification {
```
Comment (pospi) [inline zomes/rea_economic_resource/storage/src/lib.rs]: I think there's an error in this logic. `EconomicResource.UnitOfEffort` should come from the `EconomicEvent.effortQuantity.hasUnit` if set and fallback to the one from the associated `ResourceSpecification.defaultUnitOfEffort`, correct @fosterlynn?

That being the case I think we can also make this a bit more efficient by only querying the `ResourceSpecification` if the necessary fields aren't set in the `EconomicEvent`.

**Review by pospi — CHANGES_REQUESTED** [review]

File: modules/vf-graphql-holochain/resolvers/resourceSpecification.ts (hunk 1 above)
Comment (Connoropolous) [reply modules/vf-graphql-holochain/resolvers/resourceSpecification.ts:50]: would have to destructure the left side of this
`record: { defaultUnitOfResource: AddressableIdentifier }`

Comment (pospi) [reply modules/vf-graphql-holochain/resolvers/resourceSpecification.ts:50]: *shrugs* it's a minor thing :)

File: zomes/rea_economic_resource/storage/src/lib.rs (hunk 2 above)
Comment (Connoropolous) [reply zomes/rea_economic_resource/storage/src/lib.rs]: sounds right. K

Comment (fosterlynn) [reply zomes/rea_economic_resource/storage/src/lib.rs]: > EconomicResource.UnitOfEffort should come from the EconomicEvent.effortQuantity.hasUnit if set and fallback to the one from the associated ResourceSpecification.defaultUnitOfEffort, correct @fosterlynn?

Well, it is hard to imagine an event that creates a resource having an effortQuantity.  So it should take the one from the resource specification as you suggested, if it is there, when created.  I also think we need the capability to set it directly as part of creating or updating a resource, which would take precedence over the one in the resource specification.  I also don't see a big problem with doing what you suggest with the effortQuantity.hasUnit if it is not already set.  (Not sure if we want to overwrite it in that situation, but could perhaps be persuaded. It seems like a user choice, and I'm actually not sure how many user groups would want it to be a firm setting, or just a default that can be overridden on particular events.)

Comment (Connoropolous) [reply zomes/rea_economic_resource/storage/src/lib.rs]: "So it should take the one from the resource specification as you suggested, if it is there, when created." this bit we already do

**Pull request merged**